haskell/cabal.bzl

```diff
@@ -2653,34 +2653,35 @@ _use_stack = repository_rule(
 )
 
 def use_stack(stack):
-    """force given `stack` binary in all invocations of `stack_snapshot`.
+    """Force given `stack` binary in all invocations of `stack_snapshot`.
 
-    allows global override of `stack` executable, independent of call site,
+    Allows global override of `stack` executable, independent of call site,
     specifically from dependants of a `stack_snapshot`.
+    Call this in `WORKSPACE` before any use of `stack_snapshot`.
 
-    call this in `WORKSPACE` before any use of `stack_snapshot`.
+    WARNING: Use it only as a stop-gap measure if tracing discrepancies in
+    `stack` versions becomes overwhelming. Try to avoid its spooky action at
+    distance, and instead explicitly pass `stack` to macros which end up
+    calling `stack_snapshot`. It should always be possible and is a bug otherwise.
```

Comment:
I understand the sentiment, but this seems a bit out of place. `WORKSPACE` files are not declarative, instead they are sequential and imperative. This kind of behavior is very common in `WORKSPACE` files, in fact the common approach for declaring external dependencies is to use [`maybe`](https://bazel.build/rules/lib/repo/utils#maybe), which has a similar kind of behavior: The first declaration of a dependency is effectively an override of all later ones. So, this approach fits naturally into that model.
(For a more declarative model we'll have to wait for `bzlmod`.)

I'm not aware of a need to use different stack versions at different `stack_snapshot` instances. So, I'm not sure when using the `stack` attribute would be preferable.

That said, what does seem appropriate as a warning is that this should only be used in the top-level repository. Dependencies should not implicitly impose global overrides as they could conflict with other dependencies or the main project.
Perhaps it could be phrased like this: "WARNING: This will globally override the stack version used by all instances of stack_snapshot. This should only be used in the top-level repository."